cpp/src/slice2cs/Gen.cpp

```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
 }
 
 void
-Slice::CsVisitor::writeMarshalParams(const OperationPtr& op,
-                                     const MemberList& requiredParams,
-                                     const MemberList& taggedParams,
-                                     bool isReturnValue,
-                                     const string& customStream,
-                                     const string& obj)
+Slice::CsVisitor::writeMarshal(const OperationPtr& operation, bool returnType)
 {
-    const string stream = customStream.empty() ? "ostr" : customStream;
-    string ns = getNamespace(op->interface());
+    const string stream = "ostr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
-    if (requiredParams.size() > 0)
+    if (requiredMembers.size() > 0)
     {
-        if (isReturnValue)
+        if (returnType)
         {
-            bitSequenceSize = op->returnBitSequenceSize();
+            bitSequenceSize = operation->returnBitSequenceSize();
         }
         else
         {
-            bitSequenceSize = op->paramsBitSequenceSize();
+            bitSequenceSize = operation->paramsBitSequenceSize();
         }
+    }
+
+    bool write11ReturnLast = returnType && operation->hasReturnAndOut() && !members.front()->tagged() &&
+        requiredMembers.size() > 1;
+    if (write11ReturnLast)
+    {
+        _out << nl << "if (" << stream << ".Encoding != ZeroC.Ice.Encoding.V1_1)";
+        _out << sb;
+    }
 
+    bool repeat = false;
+    do
+    {
         if (bitSequenceSize > 0)
         {
             _out << nl << "var bitSequence = " << stream << ".WriteBitSequence(" << bitSequenceSize << ");";
```

Comment:
We are now declaring the `bitSequence`, for the if and the else blocks, we can move it out the if/else and just reset bitSequenceIndex when we repeat

Reply:
I think it's clearer to do the whole marshaling/unmarshaling in each block, including these redundant bitSequence variables. I initially tried to create a block just for the return value, but it gets more complicated (and actually didn't work).

Reply:
Fixed as suggested.